### PR TITLE
Add weight examples and semantics

### PR DESCRIPTION
**Pull request by ajreynol**

### CODE REVIEW

File: language/sygus.tex [hunk 1]
```diff
@@ -1428,10 +1428,10 @@ \subsection{Setting Benchmark Information}
 format used in the benchmark.
 A benchmark whose header contains the line
 \[
-\paren{\setinfokwd\mbox{ }:{\tt sygus\mbox{-}version\mbox{ }2.0}}
+\paren{\setinfokwd\mbox{ }:{\tt sygus\mbox{-}version\mbox{ }2.1}}
```
Comment (SaswatPadhi) [inline language/sygus.tex:1431]: Should the 2.0 on line 1433 below also be 2.1?

File: language/sygus.tex [hunk 2]
```diff
@@ -1781,18 +1781,55 @@ \subsubsection{Weights}
...
+to denote weighted production rules.
 The class of nullary integer symbols of the above form are
-interpreted as the sum of weights of terms used for generating the body of $f$.
+interpreted as the sum of weights of production rules used for generating the body of $f$.
+
+Formally,
+let $G$ be a grammar provided for function-to-synthesize $f$, and let $W$ be a weight keyword.
+Let $y$ be a non-terminal symbol of $G$ and let $t$ be a term appearing
+in the grouped rule listing for $y$, as described in the syntax for grammar
+definitions in Section~\ref{ssec:dec-synth-fun}.
```
Comment (SaswatPadhi) [inline language/sygus.tex]: To be consistent with other references, please use `\cref{ssec:dec-synth-fun}` here.

File: language/sygus.tex [hunk 3]
```diff
@@ -1781,18 +1781,55 @@ \subsubsection{Weights}
 \label{sec:weight-semantics}
 For each function-to-synthesize $f$ and weight keyword $W$
 as mentioned in Section~\ref{sec:weight-attributes},
```
Comment (SaswatPadhi) [inline language/sygus.tex]: To be consistent with other references, please use `\cref{ssec:weight-attributes}` here.

File: language/sygus.tex [hunk 4]
```diff
@@ -1781,18 +1781,55 @@ \subsubsection{Weights}
...
+the symbol \weightsymnf{$w$}{$f$} can be interpreted as the value 
+$k_1 + \ldots + k_n$ if it is possible to construct a sequence of terms
+$s_1, \ldots, s_n$
+with $s_1$ is the starting symbol of $G$ and $s_n = r$
+where for each $1 \leq i \leq n$, term $s_i$ is obtained from $s_{i-1}$ by
+replacing an occurrence of some $y_i$ by $t_i$
+where $y_i \mapsto_{W,k_i} t_i$ is a rule in $G$.
+
+Notice that since there may be \emph{multiple} sequences of terms that
+generates the same term $r$ above.
```
Comment (SaswatPadhi) [inline language/sygus.tex]: s/generates/generate

File: language/sygus.tex [hunk 5]
```diff
@@ -1781,18 +1781,55 @@ \subsubsection{Weights}
...
+When the body of function-to-synthesize $f$ is interpreted as term $r$,
+the symbol \weightsymnf{$w$}{$f$} can be interpreted as the value 
+$k_1 + \ldots + k_n$ if it is possible to construct a sequence of terms
+$s_1, \ldots, s_n$
+with $s_1$ is the starting symbol of $G$ and $s_n = r$
+where for each $1 \leq i \leq n$, term $s_i$ is obtained from $s_{i-1}$ by
+replacing an occurrence of some $y_i$ by $t_i$
+where $y_i \mapsto_{W,k_i} t_i$ is a rule in $G$.
+
+Notice that since there may be \emph{multiple} sequences of terms that
```
Comment (SaswatPadhi) [inline language/sygus.tex]: May be ~~since~~?

File: language/sygus.tex [hunk 6]
```diff
@@ -2476,6 +2513,63 @@ \section{Examples}%
 \end{lstlisting}
 \end{example}
 
+\begin{example}[Weights]
+Consider the following example:
+\begin{lstlisting}[language=SyGuS]
+(set-logic NIA)
+(set-feature :weights true)
+(declare-weight numX)
```
Comment (SaswatPadhi) [inline language/sygus.tex:2521]: Please also add `declare-weight` to `keywords=[6]` set in `listing-style.tex` for syntax highlighting

File: language/sygus.tex [hunk 7]
```diff
@@ -2476,6 +2513,63 @@ \section{Examples}%
...
+(set-feature :weights true)
+(declare-weight numI)
+(synth-fun f ((x Int)) Int
+   ((I Int))
+   ((I Int (0 1 x (+ x 1) (! (- I) :numI 1) (! (+ I I) :numI 2)))))
+(declare-var x Int)
+(constraint (and (> (f x) x) (<= (_ numI f) 1)))
+(check-synth)
+\end{lstlisting}
+In this example, a weight keyword $\texttt{numRec}$ has been declared,
```
Comment (SaswatPadhi) [inline language/sygus.tex]: Did you mean `numI` in place of `numRec` here?

File: language/sygus.tex [hunk 8]
```diff
@@ -1058,7 +1058,7 @@ \subsection{Declaring Sorts}
 
 \end{itemize}
 
-\subsection{Declaring Weight}
+\subsection{Declaring Weight Keywords}
```
Comment (SaswatPadhi) [inline language/sygus.tex:1061]: Please also add `declare-weight` to the reserved words in appendix A.

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
Thanks Andy.

I have a few comments, all minor though:

**Comment by ajreynol** [comment]
I've addressed the comments, the reserved words I've also updated to include the oracle commands.  Also, I've improved the 2nd example to be a bit more intuitive.

**Comment by SaswatPadhi** [comment]
Thanks! Looks great. Yay! Now we have support for weighted SyGuS :)

Merging this.

**Pull request merged**